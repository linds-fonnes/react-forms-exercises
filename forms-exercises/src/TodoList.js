import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import { v4 as uuid } from "uuid";

function TodoList() {
  const [items, setItems] = useState([]);

  const renderItems = () => {
    return (
      <ul>
        {items.map((item) => (
          <Todo item={item} removeItem={removeItem} />
        ))}
      </ul>
    );
  };

  const removeItem = (id) => {
    setItems((items) => items.filter((item) => item.id !== id));
  };

  const addItem = (item) => {
    let newItem = { ...item, id: uuid() };
    setItems((item) => [...items, newItem]);
  };

  return (
    <div className="TodoList">
      <NewTodoForm addItem={addItem} />
      {renderItems()}
    </div>
  );
}

export default TodoList;
