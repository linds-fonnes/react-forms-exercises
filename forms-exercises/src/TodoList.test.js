import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addItem(todoList, name = "walk dog") {
  const todoInput = todoList.getByLabelText("Todo:");

  fireEvent.change(todoInput, { target: { value: name } });

  const button = todoList.getByText("Add todo");
  fireEvent.click(button);
}

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("adds a new todo to list", function () {
  const todoList = render(<TodoList />);
  //should not be any todos on list yet
  expect(todoList.queryByText("x")).not.toBeInTheDocument();
  addItem(todoList);

  //todo and remove button should now be in list
  const removeBtn = todoList.getByText("x");
  expect(removeBtn).toBeInTheDocument();
  expect(todoList.queryByText("walk dog")).toBeInTheDocument();
});

it("removes a todo from list", function () {
  const todoList = render(<TodoList />);
  addItem(todoList);

  const removeBtn = todoList.getByText("x");
  fireEvent.click(removeBtn);
  expect(todoList.queryByText("walk dog")).not.toBeInTheDocument();
  expect(removeBtn).not.toBeInTheDocument();
});
