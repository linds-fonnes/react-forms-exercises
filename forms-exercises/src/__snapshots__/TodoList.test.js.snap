// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form>
      <label
        for="name"
      >
        Todo:
      </label>
      <input
        id="name"
        name="name"
        value=""
      />
      <button>
        Add todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
